### PR TITLE
[WIP] Support versionCompatibility with dependencies

### PR DESCRIPTION
**Pull request by offirgolan**
Hi there! 

I have come across many cases where I need an ember-try config that supports Ember 1.11+ with the appropriate Ember Data dependency. With this PR, I hope to solve that issue by adding support to build a config with versionCompatibility + ED. 

Without this, you would have to create a custom scenario for each ember version with the correct ED version. After a while, the ember-try config just get bigger and bigger and honestly, becomes really hard to manage. 
### Usage

``` json
"ember-addon": {
    "versionCompatibility": {
       "ember": ">1.11.0 <3.0.0",
      "withEmberData": true
    }
  }
```

**Note:** `withEmberData` can be changed to something else. Totally up for a renaming. 
### Features
- Add the corresponding latest patch release of the matching minor version of ED 
- If ember < 2.3.0, add ED to both bower and npm
### Questions
1. In the case where Ember < 1.13, there is no matching ED version. The current implementation just skips the addition of ED into the scenario. What should be done in this case?  
### Todo
- [x] Add support for Ember >= 1.13 + ED
- [ ] Support Ember < 1.13 + ED
- [ ] Add test cases
- [ ] Update docs

### CODE REVIEW

**Comment by offirgolan** [comment]
ping @kategengler @rwjblue

**Comment by offirgolan** [comment]
@rwjblue the current version of this PR holds the POC we discussed.

- Caches known bower.json and package.json files (this can be removed)
- Merges __all__ bower and npm dependencies into the scenario config if the major and minor versions match

File: data/dependencies/2.3.0-beta.1-bower.json [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+    "ember": "2.3.0",
+    "ember-cli-shims": "0.1.0",
+    "ember-cli-test-loader": "0.2.2",
+    "ember-load-initializers": "0.1.7",
+    "ember-qunit-notifications": "0.1.0",
+    "jquery": "1.11.3",
+    "loader.js": "^3.5.0"
+  },
+  "resolutions": {
+    "ember": "2.2.0"
```
Comment (kategengler) [inline data/dependencies/2.3.0-beta.1-bower.json]: Is this the right resolution for this version of ember?

File: lib/auto-scenario-config-for-ember.js [hunk 2]
```diff
@@ -2,72 +2,168 @@
...
+
+  /*
+    Add ember-data to npm dependencies
+   */
+  if (includeDependencies) {
+    beta.npm = {
+      devDependencies: {
+        'ember-data': 'components/ember-data#beta'
+      },
+      resolutions: {
```
Comment (kategengler) [inline lib/auto-scenario-config-for-ember.js]: I'm not sure npm has resolutions.

**Comment by kategengler** [comment]
Awesome! 

Some thoughts:
- Some of the cache files seem to reference ember versions different than the filename would suggest. I know these were probably fetched from ember-new-output, but its possible that wasn't always correct?

- What about going out to github for the files only if the cache version doesn't exist? (Apologies if that is in there, I may have missed it)

- Does changing all those dependencies work? I know in particular I've tried to use ember-try changing ember-cli version and haven't yet gotten that to work (I also haven't had much time to try it though). 

- How about instead of `useEmberData: true` directly in the version compatibility, we use ember-data if ember-data is in their package.json? `ember-try` already passes some info from package.json in here, and so could pass a param into this module if ember data should be used. I think it should be safe to do this from the time ember-data started versioning in sync with ember. Thoughts? cc @rwjblue 

- I wouldn't gate this work on supporting Ember < 1.13. I think it would be completely reasonable that only newer apps get this kind of nice feature, at least at first.

**Comment by offirgolan** [comment]
@kategengler 

> Some of the cache files seem to reference ember versions different than the filename would suggest. I know these were probably fetched from ember-new-output, but its possible that wasn't always correct?

This is def. an issue with ember-new-output this will need to be resolved. There are also some missing versions (2.0, 2.1, and 2.2) which will need to be added. 

> What about going out to github for the files only if the cache version doesn't exist? (Apologies if that is in there, I may have missed it)

That's the current implementation :) 

> Does changing all those dependencies work? I know in particular I've tried to use ember-try changing ember-cli version and haven't yet gotten that to work (I also haven't had much time to try it though).

There are 2 files created `data/dependencies/bower.json` and `data/dependencies/package.json` which whitelist the important dependencies that will be included

> How about instead of useEmberData: true directly in the version compatibility, we use ember-data if ember-data is in their package.json? 

I was thinking of something like:

```json
"ember-addon": {
    "versionCompatibility": {
       "ember": ">1.11.0 <3.0.0",
       "includeDependencies": true,
      "exclude": {
        "bower": [ "ember-data" ],
        "npm": [ "ember-data" ]
      }
    }
  }
```

> I wouldn't gate this work on supporting Ember < 1.13. I think it would be completely reasonable that only newer apps get this kind of nice feature, at least at first.

As per my discussion w/ @rwjblue, this will only work for Ember >= 1.13 which I think is totally reasonable as well.

**Comment by kategengler** [comment]
Ah, what I was thinking is that we opt anybody who has opted into
versionCompatibility and has ember-data as a dependency into also varying
the other dependencies. I think it makes sense because the likelihood is
that anybody using Ember 2.1 is using Ember-Data 2.1, so addons will want
to know if they work with the setup that goes with that version of Ember
(the whole point of this PR, I think). Maybe with a way to opt out, just in
case of unforeseen problems with it.

**Comment by offirgolan** [comment]
@kategengler Im really sorry for the confusion. After the initial design of only including ember-data, @rwjblue mentioned that there are other dependencies that have to come into play (thats where the need for `ember-new-output` comes into play). Im not really sure which dependencies are actually needed so that is something that we should discuss. Including / excluding ED from the dependency list is an easy task but I think we should first define some clear guidelines. Some input from the both of you on how to proceed would be ideal 😄

**Comment by kategengler** [comment]
No need to be sorry. I'm probably not being clear. I would like @rwjblue's
input on this idea as well:

Without any configuration by the end user, I'd like to have the other
dependencies changing by default, giving them a way to opt out.
Without any configuration by the end user, I'd like to have ember-data
changing if they have ember-data in their package.json (ember-try can pass
in a param to indicate this).

@rwjblue will be better able to say which dependencies should change. I
know that right now, changing ember-cli version between scenarios does not
work in ember-try.

**Comment by Turbo87** [comment]
closing due to inactivity

**Pull request closed without merging**